src/NUnitFramework/framework/Internal/PreFilter.cs

```diff
@@ -0,0 +1,208 @@
+
+        private List<FilterElement> _filters = new List<FilterElement>();
+
+        /// <summary>
+        /// Return a new PreFilter, without elements, which is considered
+        /// empty and always matches.
+        /// </summary>
+        public static PreFilter Empty
+        {
+            get { return new PreFilter(); }
+        }
+
+        /// <summary>
+        /// Return true if the filer is empty, in which case it
+        /// always succeeds. Technically, this is just a filter and
+        /// you can add elements but it's best to use Empty when
+        /// you need an empty filter and new when you plan to add.
+        /// </summary>
+        public bool IsEmpty
+        {
+            get { return _filters.Count == 0; }
+        }
+
+        /// <summary>
+        /// Add a new filter element to the filter
+        /// </summary>
+        /// <param name="filterText"></param>
+        public void Add(string filterText)
+        {
+            _filters.Add(new FilterElement(filterText));
+        }
+
+        /// <summary>
+        /// Use the filter on a Type, returning true if the type matches the filter
+        /// and should therefore be included in the discovery process.
+        /// </summary>
+        public bool Match(Type type)
+        {
+            if (IsEmpty)
+                return true;
+
+            string typeName = type.FullName;
+
+            foreach (FilterElement filter in _filters)
+            {
+                if (filter.Match(type))
+                    return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Use the filter on a Type, returning true if the type matches the filter
+        /// and should therefore be included in the discovery process.
+        /// </summary>
+        public bool Match(Type type, MethodInfo method)
+        {
+            if (IsEmpty)
+                return true;
+
+            int filterCount = 0;
+            foreach (FilterElement filter in _filters)
+            {
+                if (type.FullName == filter.ClassName)
+                {
+                    ++filterCount;
+                    if (method.Name == filter.MethodName)
+                        return true;
+                }
+            }
+
+            // If no match, we succeed only if there were no method filters for this fixture
+            return filterCount == 0;
+        }
+
+        #region Nested FilterElementType Enumeration
+
+        private enum FilterElementType
+        {
+            Unknown,
+            Namespace,
+            Fixture,
+            Method
+        }
+
+        #endregion
+
+        #region Nested FilterElement Class
+
+        private class FilterElement
+        {
+            public FilterElementType ElementType = FilterElementType.Unknown;
+            public string Text;
+            public string ClassName;
+            public string MethodName;
```

Comment:
Can these all be `private` and the three strings `readonly` (and the class `sealed` 😀)?

Reply:
Why? It's a private nested class that exists solely at the convenience of the `PreFilter` class. Only reason it's not a struct is because of the rules for struct constructors.

Reply:
@CharliePoole Because all those things make reading comprehension faster by cutting down on what-ifs that must be mentally tracked.
I thought about suggesting `struct`, but the problem is you'd have to copy it back into the `_filters` list each time it's used since it's mutable or else `ElementType` will just stay 0 the whole time.

Reply:
@CharliePoole Because all those things make reading comprehension faster by cutting down on what-ifs that must be mentally tracked. Scanning the mutability of the fields is for me lot like scanning the UPPER_CASE constant parameters is for you.

I thought about suggesting `struct`, but the problem is you'd have to copy it back into the `_filters` list each time it's used since it's mutable or else `ElementType` will just stay 0 the whole time.